src/pipelinedefinition.cpp

```diff
@@ -85,14 +87,70 @@ Status PipelineDefinition::validate(ModelManager& manager) {
+            void* customNodeLibraryInternalManager = nullptr;
+            auto it = nodeResources.find(nodeInfo.nodeName);
+            bool internalManagerExists = (it != nodeResources.end());
+            if (internalManagerExists) {
+                customNodeLibraryInternalManager = it->second;
+            }
+            auto params = createCustomNodeParamArray(nodeInfo.parameters).get();
+            if (!nodeInfo.library.isValid()) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Pipeline: {} node: {} refers to invalid library", pipelineName, nodeInfo.nodeName);
+                return StatusCode::PIPELINE_DEFINITION_INVALID_NODE_LIBRARY;
+            }
+            auto status = nodeInfo.library.initialize(&customNodeLibraryInternalManager, params, nodeInfo.parameters.size());
+            if (status != 0) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Initialization of library with base path: {} failed", nodeInfo.library.basePath);
+                return StatusCode::NODE_LIBRARY_INITIALIZE_FAILED;
+            }
+            if (!internalManagerExists) {
+                nodeResources.insert({nodeInfo.nodeName, customNodeLibraryInternalManager});
+            }
+        }
+    }
+    return StatusCode::OK;
+}
+
+std::vector<NodeInfo> PipelineDefinition::calculateNodeInfosDiff(const std::vector<NodeInfo>& nodeInfos) {
+    std::vector<NodeInfo> diff;
+    for (const auto& nodeInfo : this->nodeInfos) {
+        auto it = std::find_if(nodeInfos.begin(), nodeInfos.end(),
+            [&nodeInfo](const auto& x) { return x.nodeName == nodeInfo.nodeName; });
+        if (it == nodeInfos.end()) {
+            diff.push_back(nodeInfo);
+        }
+    }
+    return diff;
+}
+
+void PipelineDefinition::deinitializeNodeResources(const std::vector<NodeInfo>& nodeInfosDiff) {
+    for (const auto& nodeInfo : nodeInfosDiff) {
+        if (nodeInfo.kind == NodeKind::CUSTOM) {
+            auto it = nodeResources.find(nodeInfo.nodeName);
+            if (it == nodeResources.end()) {
+                continue;
+            }
+            void* customNodeLibraryInternalManager = it->second;
+            auto status = nodeInfo.library.deinitialize(customNodeLibraryInternalManager);
+            if (status != 0) {
+                SPDLOG_LOGGER_ERROR(modelmanager_logger, "Deinitialization of library with base path: {} failed", nodeInfo.library.basePath);
+            }
+            nodeResources.erase(nodeInfo.nodeName);
+        }
+    }
+}
+
 Status PipelineDefinition::reload(ModelManager& manager, const std::vector<NodeInfo>&& nodeInfos, const pipeline_connections_t&& connections) {
     // block creating new unloadGuards
     this->status.handle(ReloadEvent());
     resetSubscriptions(manager);
     while (requestsHandlesCounter > 0) {
         std::this_thread::sleep_for(std::chrono::microseconds(1));
     }
-
+    deinitializeNodeResources(calculateNodeInfosDiff(nodeInfos));
```

Comment:
This shouldn't be here. Idea is to let implementation of custom node "initialize" method decide if any blocking of predict requests is needed.

With first creation we call initialize with void** nodeResource, and *nodeResource should be nullptr.
With reloads we call initialize with void** nodeResource but this time *nodeResource should point to already loaded resource/manager.

It seems we need test to prove that (ie -> count initalize calls and with createDefinition + reload) it should be 2. Consider that when changing eg. model nireq setting this would mean that PD is not available, and what's worse undergoing predictions would be wasted.

Reply:
This would only affect nodes that are not in pipeline definition anymore. calculateNodeInfosDiff returns std::void<NodeInfo> with nodeInfos that are not in new vector provieded via PipelineDefinition::reload function argument, but were in latest nodeInfos, which at the function call moment are saved as PipelineDefinition member.